### PR TITLE
add InfrastructureTopology configuration support

### PR DESCRIPTION
**Pull request by hanqiuzh**
### Updates
- added `infrastructureAvailabilityPolicy` to hostedcluster & hostedcontrolplane
   - default ingresscontroller: spec.replicas=1/2 for singleReplica/HA
   - infrastructure: status.infrastructureTopology=hcp.spec.infrastructureAvailabilityPolicy
- updated openshift/api to release-4.8
- updated related packages 
   - Note: k8s.io/kube-aggregator is not upgraded as it needs more work

### Problems
- infrastructureAvailabilityPolicy updates after cluster is ready won't be propagated to the cluster (not sure if it's needed):
   - configmap of user manifests are updated, but they are applied in `manifest-bootstrapper` pod, which doesn't re-apply when configs updated
   - infrastructure status controller [setup function only read the configmap once](https://github.com/openshift/hypershift/blob/2f090b29cf8b66368ebe167e616b2a4d31da231e/hosted-cluster-config-operator/controllers/infrastatus/setup.go#L30)
   - only can be updated if we delete:
      - `manifests-bootstrapper`, `control-plane-operator` (re-apply default ingresscontroller)
      - `hosted-cluster-config-operator` (reload infrastatus controller)

### Tests
- [x] if hostedcluster's infrastructureAvailabilityPolicy = singlereplica, console/router-default/thanos-querier has one replica
- [x] if hostedcluster's infrastructureAvailabilityPolicy = highlyavailable, console/router-default/thanos-querier has 2+ replicas
- [x] ~~TODO: e2e autoscaling cover 1 node test case~~ autoscaling will work for 1->3 nodes, but cannot do 3->1 nodes as the [csi controller](https://github.com/openshift/aws-efs-csi-driver-operator/blob/625e39dc583adc79a1edbb0d6691d201eeaf21e4/vendor/github.com/openshift/library-go/pkg/operator/csi/csidrivercontrollerservicecontroller/helpers.go#L85-L92) will have 2 pods when we have more than 2 nodes.

### CODE REVIEW

File: api/v1alpha1/hosted_controlplane.go [hunk 1]
```diff
@@ -52,6 +52,12 @@ type HostedControlPlaneSpec struct {
 	// +optional
 	ControllerAvailabilityPolicy AvailabilityPolicy `json:"controllerAvailabilityPolicy,omitempty"`
 
+	// InfrastructureAvailabilityPolicy specifies whether to run infrastructure services that
```
Comment (enxebre) [inline api/v1alpha1/hosted_controlplane.go:55]: may be mention this affects to operators running on the guest cluster

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 2]
```diff
@@ -1810,6 +1815,13 @@ func (r *HostedControlPlaneReconciler) generateControlPlaneManifests(ctx context
 	params.ImageRegistryHTTPSecret = generateImageRegistrySecret()
 	params.APIAvailabilityPolicy = render.SingleReplica
 	params.ControllerAvailabilityPolicy = render.SingleReplica
+	if hcp.Spec.ControllerAvailabilityPolicy == hyperv1.HighlyAvailable {
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:1818]: was there any template actually using this? Now this is propagated to the guest cluster infra CR but Is there any component in the guest cluster that will actually use this now? If not we should probably add a comment here to clarify that.
cc @csrwng

**Comment by enxebre** [comment]
thanks @hanqiuzh! overall this looks great to me! PTAL @ironcladlou @csrwng @sjenning

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 2 above)
Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:1818]: it was not used anywhere, same as `APIAvailabilityPolicy`, and I didn't find any place using that variable. Is it some dead code? 

I added `ControllerAvailabilityPolicy` into the `infrastructure.status.controlPlaneTopology`, although it won't affect anything 😂

File: control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml [hunk 3]
```diff
@@ -16,6 +16,8 @@ status:
   etcdDiscoveryDomain: {{ .BaseDomain }}
   infrastructureName: "{{ .InfraID }}"
   platform: {{ .PlatformType }}
+  controlPlaneTopology: {{ .ControllerAvailabilityPolicy }}
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml:19]: controlPlaneTopology should be `External` (not dependent on ControllerAvailabilityPolicy)

Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml:19]: `External` is added in 4.9? currently this branch is using 4.8, so doesn't have that field 😢

Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml:19]: Just leave it unset for now then

Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml:19]: https://github.com/openshift/api/blob/release-4.8/config/v1/types_infrastructure.go#L86
just tried. seems the field doesn't omitempty 😵 and it cannot be applied with blank?
```
2021-09-07T18:11:04.475Z	INFO	hypershift-operator.InfraStatus	Updating infrastructure status
2021-09-07T18:11:04.481Z	ERROR	controller-runtime.manager.controller.infra-status	Reconciler error	{"name": "cluster", "namespace": "", "error": "Infrastructure.config.openshift.io \"cluster\" is invalid: status.controlPlaneTopology: Unsupported value: \"\": supported values: \"HighlyAvailable\", \"SingleReplica\""}
```

shall we always give it a default value? e.g. HighlyAvailable/SingleReplica

Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-infrastructure-02-config.yaml:19]: HighlyAvailable is fine

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 2 above)
Comment (enxebre) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: So can we completely remove `params.ControllerAvailabilityPolicy`? cc @csrwng

**Comment by enxebre** [comment]
lgtm other than this question https://github.com/openshift/hypershift/pull/454/commits/e9d184093c8c5c8e7561cb1f415279ca56af4076#r704341970

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 2 above)
Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: Agree, I don't think it's used in the remaining templates

Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: cool `params.ControllerAvailabilityPolicy` is removed, and it works 😂

**Pull request merged**